src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -2600,5 +2623,273 @@ private void Remove(int startIndex, int count, out StringBuilder chunk, out int
+            /// <remarks>This is intended to be called only by compiler-generated code. Arguments are not validated as they'd otherwise be for members intended to be used directly.</remarks>
+            public AppendInterpolatedStringHandler(int literalLength, int formattedCount, StringBuilder stringBuilder, IFormatProvider? provider)
+            {
+                _stringBuilder = stringBuilder;
+                _provider = provider;
+                _hasCustomFormatter = provider is not null && DefaultInterpolatedStringHandler.HasCustomFormatter(provider);
+            }
+
+            /// <summary>Writes the specified string to the handler.</summary>
+            /// <param name="value">The string to write.</param>
+            public void AppendLiteral(string value) => _stringBuilder.Append(value);
+
+            #region AppendFormatted
+            // Design note:
+            // This provides the same set of overloads and semantics as DefaultInterpolatedStringHandler.
+
+            #region AppendFormatted T
+            /// <summary>Writes the specified value to the handler.</summary>
+            /// <param name="value">The value to write.</param>
+            public void AppendFormatted<T>(T value) => AppendFormatted(value, format: null);
+
+            /// <summary>Writes the specified value to the handler.</summary>
+            /// <param name="value">The value to write.</param>
+            /// <param name="format">The format string.</param>
+            public void AppendFormatted<T>(T value, string? format)
+            {
+                if (_hasCustomFormatter)
+                {
+                    // If there's a custom formatter, always use it.
+                    AppendCustomFormatter(value, format);
+                }
+                else if (value is IFormattable)
+                {
+                    // Check first for IFormattable, even though we'll prefer to use ISpanFormattable, as the latter
+                    // requires the former.  For value types, it won't matter as the type checks devolve into
+                    // JIT-time constants.  For reference types, they're more likely to implement IFormattable
+                    // than they are to implement ISpanFormattable: if they don't implement either, we save an
+                    // interface check over first checking for ISpanFormattable and then for IFormattable, and
+                    // if it only implements IFormattable, we come out even: only if it implements both do we
+                    // end up paying for an extra interface check.
+
+                    if (value is ISpanFormattable)
+                    {
+                        Span<char> destination = _stringBuilder.RemainingCurrentChunk;
+                        if (((ISpanFormattable)value).TryFormat(destination, out int charsWritten, format, _provider)) // constrained call avoiding boxing for value types
```

Comment:
Nit: Since `ISpanFormattable` is now public, this could expose uninitialized memory to the caller, as `StringBuilder` uses `GC.AllocateUninitializedArray` under the covers. In the worst case, this could allow me to write a buggy `ISpanFormattable` implementation as follows.

```cs
public class MyClass : ISpanFormattable
{
    public bool TryFormat(Span<char> destination, out int charsWritten, ...)
    {
        charsWritten = destination.Length;
        return true;
    }
}
```

If I then pass an instance of that type into the new `StringBuilder.Append` overloads, the resulting `StringBuilder` will contain data populated from uninitialized memory, even though the caller (me) never wrote unsafe code or used any unsafe-equivalent API.

Now, I don't think this is a problem in practice, as nobody would ever write an `ISpanFormattable` like that. But it's an example of where we're starting to play a little fast and loose with type \& memory safety, and we may need to start considering how this affects more common APIs like `StringBuilder` (and potentially `MemoryStream`, if we decide to make these same changes there). Are these patterns deserving of mention on MSDN? Is it even worth spending any time on this at all? Maybe in the extreme it's not much different than what we already have in the `Stream.Read[Async]` base implementations, which use the array pool, and I'm being overly paranoid.

Reply:
Thanks.

> we're starting to play a little fast and loose with type & memory safety

I guess I don't see how this has anything to do with type / memory safety.  Nothing promises that the destination buffer is zeroes. It might result in a gibberish string, but it's not breaking type safety, not causing the runtime to go off and do wacky things (and if it is, that's a bug in the runtime to be fixed), not causing the runtime to misinterpret one type for another, etc.